streams/src/main/java/org/apache/kafka/streams/state/internals/ReadonlyPartiallyDeserializedSegmentValue.java

```diff
@@ -0,0 +1,210 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.query.ResultOrder;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+
+
+final class ReadonlyPartiallyDeserializedSegmentValue {
+
+    private static final int TIMESTAMP_SIZE = 8;
+    private static final int VALUE_SIZE = 4;
+    private byte[] segmentValue;
+    private long nextTimestamp;
+    private long minTimestamp;
+
+    private int deserIndex = -1; // index up through which this segment has been deserialized (inclusive)
+
+    private Map<Integer, Integer> cumulativeValueSizes;
+
+    private int valuesStartingIndex = -1; // the index of the first value in the segment (but the last one in the list)
+    private Map<Integer, TimestampAndValueSize> unpackedTimestampAndValueSizes = new HashMap<>();
+    private int recordNumber = -1; // number of segment records
+
+
+    ReadonlyPartiallyDeserializedSegmentValue(final byte[] segmentValue) {
+        this.segmentValue = segmentValue;
+        this.nextTimestamp =
+                RocksDBVersionedStoreSegmentValueFormatter.getNextTimestamp(segmentValue);
+        this.minTimestamp =
+                RocksDBVersionedStoreSegmentValueFormatter.getMinTimestamp(segmentValue);
+        resetDeserHelpers();
+    }
+
+
+    public long minTimestamp() {
+        return minTimestamp;
+    }
+
+    public long nextTimestamp() {
+        return nextTimestamp;
+    }
+
+    public byte[] serialize() {
+        return segmentValue;
+    }
+
+
+    public RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult find(
+            final long fromTime, final long toTime, final ResultOrder order, final int index) {
+
+        // this segment does not have any record in query specified time range
+        if (toTime < minTimestamp || fromTime > nextTimestamp) {
+            return null;
+        }
+
+        final boolean isAscending = order.equals(ResultOrder.ASCENDING);
+
+        if (isAscending && valuesStartingIndex == -1) {
+            findValuesStartingIndex();
+            deserIndex = recordNumber;
+        }
+
+        long currTimestamp = -1;
+        long currNextTimestamp = -1;
+        int currIndex = initializeCurrentIndex(index, isAscending);
+        int cumValueSize = initializeCumValueSize(index, currIndex, isAscending);
```

Comment:
> Not sure why we need to init `cumValueSize`? It seem below we never read the value set here?

We must do that. Look at line#100 (`cumValueSize += Math.max(currValueSize, 0);`). 
In fact, when we have everything ready in the cache (`unpackedTimestampAndValueSizes`), we do not need it. Otherwise, we need the former computed cumValueSize to be able to move to the correct position in the segment and deserialize the next one.

Reply:
line#100 is `unpackedTimestampAndValueSizes.put(currIndex, new TimestampAndValueSize(currTimestamp, cumValueSize));` in this PR. I only see `cumValueSize += Math.max(currValueSize, 0);` in line#112